models/t_parse.go

```diff
@@ -0,0 +1,53 @@
+package models
+
+import (
+	"net"
+
+	"github.com/pkg/errors"
+)
+
+// PopulateFromString populates a RecordConfig given a type and string.
+// Many providers give all the parameters of a resource record in one big
+// string (all the parameters of an MX, SRV, CAA, etc). Rather than have
+// each provider rewrite this code many times, here's a helper function to use.
+//
+// At this time, the idiom is to panic rather than continue with potentially
+// misunderstood data. We do this panic() at the provider level.
+// Therefore the typical calling sequence is:
+//     if err := rc.PopulateFromString(rtype, value, origin); err != nil {
+//         panic(errors.Wrap(err, "unparsable record received from provider"))
+//     }
+func (r *RecordConfig) PopulateFromString(rtype, contents, origin string) error {
+	if r.Type != "" && r.Type != rtype {
+		panic(errors.Errorf("assertion failed: rtype already set (%s) (%s)", rtype, r.Type))
```

Comment:
I'm even more confused by panic here. We return an error, but also panic sometimes? What is going on?